pulp_2to3_migration/app/plugin/deb/pulp_2to3_models.py

```diff
@@ -0,0 +1,157 @@
+        'installed_size': 'Installed-Size',
+        'maintainer': 'Maintainer',
+        'original_maintainer': 'Original-Maintainer',
+        'architecture': 'Architecture',
+        'replaces': 'Replaces',
+        'provides': 'Provides',
+        'depends': 'Depends',
+        'pre_depends': 'Pre-Depends',
+        'recommends': 'Recommends',
+        'suggests': 'Suggests',
+        'enhances': 'Enhances',
+        'conflicts': 'Conflicts',
+        'breaks': 'Breaks',
+        'description': 'Description',
+        'multi_arch': 'Multi-Arch',
+        'homepage': 'Homepage',
+        'section': 'Section',
+        'priority': 'Priority',
+    }
+
+    class Meta:
+        default_related_name = 'deb_detail_model'
+        unique_together = ('name', 'version', 'architecture', 'checksumtype', 'checksum', 'pulp2content')
+
+    @property
+    def expected_digests(self):
+        """Return expected digests."""
+        return {self.checksumtype: self.checksum}
+
+    @property
+    def expected_size(self):
+        """Return expected size."""
+        return self.size
+
+    @property
+    def relative_path_for_content_artifact(self):
+        """Return relative path."""
+        return self.filename
+
+    @classmethod
+    def pre_migrate_content_detail(cls, content_batch):
+        """
+        Pre-migrate Pulp 2 content with all the fields needed to create a Pulp 3 Content
+
+        Args:
+             content_batch(list of Pulp2Content): pre-migrated generic data for Pulp 2 content.
+        """
+        pulp2_unit_fields = set([
+            'checksumtype',
+            'checksum',
+            'filename',
+            'size',
+            'control_fields',
+        ])
+        pulp2_id_obj_map = {pulp2content.pulp2_id: pulp2content for pulp2content in content_batch}
+        pulp2_ids = pulp2_id_obj_map.keys()
+        pulp2_content_batch = pulp2_models.DebPackage.objects.filter(id__in=pulp2_ids).as_pymongo().only(*pulp2_unit_fields)
+
+        pulp2deb_to_save = []
+        for deb in pulp2_content_batch:
+            pre_migrate_fields = {k: deb['control_fields'][v] for k, v in cls.control_field_map.items() if v in deb['control_fields'].keys()}
+            keys = deb['control_fields'].keys() - cls.control_field_map.values()
+            extra_control_fields = {k: deb['control_fields'][k] for k in keys}
+            pulp2deb_to_save.append(
+                Pulp2DebPackage(
+                    checksumtype=deb['checksumtype'],
+                    checksum=deb['checksum'],
+                    filename=deb['filename'],
+                    extra_control_fields=json.dumps(extra_control_fields),
+                    pulp2content=pulp2_id_obj_map[deb['_id']],
+                    size=deb['size'],
+                    **pre_migrate_fields,
+                )
+            )
+
+        cls.objects.bulk_create(pulp2deb_to_save, ignore_conflicts=True, batch_size=DEFAULT_BATCH_SIZE)
+
+    @classmethod
+    def create_pulp3_content(self):
```

Comment:
As far as I can tell everything works fine until `create_pulp3_content` is called.
However within this function I can't seem to access any of the fields of the pre migrated units being passed to it.
Every field access (e.g.: `self.checksum`) appears to return a `DeferredAttribute` object, rather than a string value.

Reply:
I added a `self.checksum` call and printed the result to log output, and got:
```
<django.db.models.query_utils.DeferredAttribute object at 0x7f056bb8a8d0>
```

Reply:
Within the DB things look right to me, e.g.:
```
pulp=# SELECT checksum FROM pulp_2to3_migration_pulp2debpackage WHERE filename='nvidia-container-toolkit_1.0.5-1_amd64.deb';
-[ RECORD 1 ]--------------------------------------------------------------
checksum | 9ffe51f7520528ef9215a556abb641a16dbbc1e8d44b8ea2c413e0f711de21ff
```

Reply:
@quba42 The bug is actually right on this very line :)

You've defined this method as a classmethod, which passes the class through the ```self``` parameter.  Therefore when you access attributes on ```self```, they exist, but the value you get back is the type they exist as on the class, as opposed to the type you'd get when accessing an instance.

It should just be a normal method, not a classmethod.  It might have used to have been a classmethod though, I can't remember.

Reply:
That has done it! I was now able to successfully migrate Debian packages from Pulp2 to Pulp3!
There is certainly some more work to be done, but the rest of it should be a lot more straight forward.
This very much concludes an important initial milestone!